File: src_matlab/+nmpccodegen/+controller/+obstacles/Rectangular.m
classdef Rectangular < nmpccodegen.controller.obstacles.Obstacle
    %RECTANGULAR A simple rectangular obstacle.
    %  A simple two-dimensional rectangular obstacle.
    
    properties
        width % The width of the rectangular.
        height % The height of the rectangular.
        center_coordinates % The center coordinates of the rectangular.
    end
    
    methods
        function obj = Rectangular(center_coordinates,width,height,model)
            % - width = The width of the rectangular.
            % - height = The height of the rectangular.
            % - center_coordinates = The center coordinates of the rectangular.
            % - model = Model of the controlled system.
            obj@nmpccodegen.controller.obstacles.Obstacle(model)
            obj.center_coordinates=center_coordinates;
            obj.width=width;
            obj.height=height;
        end
        function cost = evaluate_coordinate_state_cost(obj,coordinates_state)
            % Evaluate the cost of the obstacle at a particular location.
            [x_up,x_down,y_up,y_down] = get_corner_coordinates(obj);
            a = ([-1. 0.; 1., 0. ;0., -1.; 0., 1.])';
            b = [x_up; -x_down; y_up; -y_down];
            
            cost =  nmpccodegen.controller.obstacles.Polyhedral(a, b, obj.model).evaluate_coordinate_state_cost(coordinates_state);
        end
        function plot(obj)
            % Plots the rectangular obstacle to the active figure.
            [x_up,x_down,y_up,y_down] = get_corner_coordinates(obj);
            plot([x_down x_up],[y_down y_down],'linewidth', 2,'Color', 'black');
            plot([x_down x_up],[y_up y_up],'linewidth', 2,'Color', 'black');
            plot([x_up x_up],[y_up y_down],'linewidth', 2,'Color', 'black');
            plot([x_down x_down],[y_down y_up],'linewidth', 2,'Color', 'black');
            plot(y_up)
        end
    end
    methods (Access = private)
        function [x_up,x_down,y_up,y_down] = get_corner_coordinates(obj)
            % Internal function used to get the coordinates of the corners
            % of the rectangular.
            x_up = obj.center_coordinates(1) + obj.width / 2;
            x_down = obj.center_coordinates(1) - obj.width / 2;
            y_up = obj.center_coordinates(2) + obj.height / 2;
            y_down = obj.center_coordinates(2) - obj.height / 2;
        end
    end
    
end

File: old_code/src_python/nmpccodegen/controller/obstacles/rectangular.py
from .obstacle import  Obstacle
from .polyhedral import Polyhedral
import casadi as cd
import matplotlib as plt
import numpy as np

class Rectangular(Obstacle):
    def __init__(self,model,center_coordinates,width,height):
        """ construct obstable of form a[i,:]^Tb +b , for all i """
        super(Rectangular, self).__init__(model)

        self._center_coordinates=center_coordinates
        self._width = width
        self._height = height

    def evaluate_coordinate_state_cost(self,coordinates_state):
        x_up = self._center_coordinates[0] + self._width / 2
        x_down = self._center_coordinates[0] - self._width / 2
        y_up = self._center_coordinates[1] + self._height / 2
        y_down = self._center_coordinates[1] - self._height / 2

        a = np.matrix([[-1., 0.], [1., 0.], [0., -1.], [0., 1.]]).T
        b = np.array([x_up, -x_down, y_up, -y_down])
        return Polyhedral(self._model,a, b).evaluate_coordinate_state_cost(coordinates_state)
    def plot(self):
        rectangular = plt.patches.Rectangle((self._center_coordinates[0]- self._width / 2,self._center_coordinates[1]- self._height / 2),self._width,self._height,fill=False)
        ax = plt.pyplot.gca()
        ax.add_patch(rectangular)

    @property
    def width(self):
        return self._width
    @property
    def height(self):
        return self._height
    @property
    def center_coordinates(self):
        return self._center_coordinates